text/0249-v2-experiments.md

```diff
@@ -0,0 +1,127 @@
+
+- Upon release of v2.0 to GA, all experimental modules will be automatically marked as stable (in fact, the metadata in
+  the code regarding stability can be removed, since everything will be stable). This means that the opportunity for the
+  team and for the community to influence the shape of existing experimental APIs is between now and then. Once we go
+  GA, we have to live with our design decisions.
+- Any APIs that cannot be graduated with confidence in time for the v2 release will be deprecated in 1.x, removed from
+  the v2 branch, and moved to the preview repository. This option should be minimized, since some customers may have
+  deployed production stacks that depend on these APIs.
+- Create a new repository called aws-cdk-previews that will be the new home for experimental development and developer
+  preview releases. The packages in this repository will be vended in the same way that 1.x packages are vended today,
+  in a variety of languages, with metadata that draws a line between new experiments and more mature developer preview
+  packages. This repo will take a dependency on the core CDK package just like any other independent library. This
+  avoids the complications associated with experimenting on a fork or a branch of the main repository. The repository
+  will be clearly marked as experimental.
+- When a package in the preview repository has reached maturity and is ready for GA, it will be copied into the main
+  repo and made a part of CDK 2. The module in the preview repository will then be marked as deprecated with clear
+  documentation to point users to the new stable version in the main repository. The lifecycle of experimental modules
+  in the preview repository will be exactly the same as it is currently in 1.x, with the same metadata markers in the
+  code.
+- The contributor guide for the CDK will be modified so that users know where to introduce new experimental code.
+- If it becomes absolutely necessary to add a new feature to v2 that would be a breaking change for users, the feature
+  will be hidden behind a feature flag. The code path will not be accessible without a specific opt-in by the user.
+
+# Drawbacks
+
+- In several places in the current 1.x CDK branch, stable modules refer to or contain experimental code. This will no
+  longer be possible after the release of v2.
+- When we graduate a module, users who have participated in the Developer Preview will have to update their imports to
+  make use of the stable version in the main repository.
+- For developers who are working on experimental code that has complex dependencies with the core and with other
+  modules, development will be more complicated.
+- In some cases, experimentation will not be possible. New features introduced to the main repository will have to be
+  vetted much more carefully before they are merged, since any public API released as part of v2 will have to be
+  supported for the lifetime of the major version.
```

Comment:
Another downside: third-party libraries effectively cannot use the experimental modules (because of the dependency problem that mono-CDK solves for stable modules).

Reply:
Why not? There are several 3rd party repositories out there like Solutions Constructs that reference experimental packages.

Reply:
From what I know, all third-party packages are pinned to specific versions of the CDK libraries, which prevents their consumers from upgrading. This is exactly the problem that mono-CDK was meant to solve. I feel like we're re-introducing it here, just for the experimental modules.

Reply:
Looks like solutions constructs upgrades whenever we release a new version. https://github.com/awslabs/aws-solutions-constructs/blob/master/source/patterns/%40aws-solutions-constructs/aws-apigateway-sqs/package.json

Reply:
I'm not sure what your argument here is @ericzbeard . Are you saying that there is no need for mono-CDK, and that the current third-party ecosystem works fine?

Reply:
No, just providing a data point. We can't depend on an external group to consistently upgrading every time we do a release. If they stopped, or pinned to a certain version, then their users would be stuck on an old version of CDK.

Reply:
Exactly. Which leads to my comment, which is: "If we release every experimental module as a separate package, we will cause exactly the same pain points the third-party libraries are going through today, just limited to the experimental modules", which is another downside of this approach (which is the section I was commenting on).

Reply:
I agree this is a big problem. If nobody can use the experimental modules, they are effectively useless. It's important to not lose sight of *why* we have experimental APIs in the first place (that is - to enable early adopters to provide us with valuable feedback before it's too late).